File: backend/app/api/v1/endpoints/roles.py
from typing import Any, List
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from tortoise.exceptions import DoesNotExist, IntegrityError

from app.api import deps
from app.models.user import Role, Permission, User
from app.schemas.user import Role as RoleSchema, RoleCreate
from app.schemas.response import Response, PageData, success

router = APIRouter()


class RoleUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class RolePermissionsUpdate(BaseModel):
    permissions: List[str]  # List of permission codes


@router.get("/", response_model=Response[PageData[RoleSchema]])
async def read_roles(
    page: int = 1,
    page_size: int = 50,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Retrieve roles.
    """
    total = await Role.all().count()
    skip = (page - 1) * page_size
    roles = await Role.all().offset(skip).limit(page_size).prefetch_related("permissions")
    
    return success(data={
        "items": roles,
        "total": total,
        "page": page,
        "page_size": page_size,
    })


@router.post("/", response_model=Response[RoleSchema])
async def create_role(
    *,
    role_in: RoleCreate,
    current_user: User = Depends(deps.PermissionChecker("user:manage")),
) -> Any:
    """
    Create new role.
    """
    try:
        role = await Role.create(
            name=role_in.name,
            description=role_in.description,
            is_system_role=False,
        )
        
        # Add permissions
        if role_in.permissions:
            for perm_code in role_in.permissions:
                try:
                    perm = await Permission.get(code=perm_code)
                    await role.permissions.add(perm)
                except DoesNotExist:
                    pass  # Skip non-existent permissions
        
        # Reload with permissions
        role = await Role.get(id=role.id).prefetch_related("permissions")
        return success(data=role, msg="Role created successfully")
    except IntegrityError:
        raise HTTPException(
            status_code=400,
            detail="Role with this name already exists",
        )


@router.get("/{role_id}", response_model=Response[RoleSchema])
async def read_role(
    role_id: UUID,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Get role by ID.
    """
    try:
        role = await Role.get(id=role_id).prefetch_related("permissions")
        return success(data=role)
    except DoesNotExist:
        raise HTTPException(
            status_code=404,
            detail="Role not found",
        )


@router.put("/{role_id}", response_model=Response[RoleSchema])
async def update_role(
    *,
    role_id: UUID,
    role_in: RoleUpdate,
    current_user: User = Depends(deps.PermissionChecker("user:manage")),
) -> Any:
    """
    Update a role.
    """
    try:
        role = await Role.get(id=role_id)
    except DoesNotExist:
        raise HTTPException(
            status_code=404,
            detail="Role not found",
        )
    
    if role.is_system_role:
        raise HTTPException(
            status_code=400,
            detail="System roles cannot be modified",
        )
    
    # Check if name is being changed and if it conflicts
    if role_in.name and role_in.name != role.name:
        existing = await Role.filter(name=role_in.name).first()
        if existing:
            raise HTTPException(
                status_code=400,
                detail="Role with this name already exists",
            )
        role.name = role_in.name
    
    if role_in.description is not None:
        role.description = role_in.description
    
    await role.save()
    
    role = await Role.get(id=role_id).prefetch_related("permissions")
    return success(data=role, msg="Role updated successfully")


@router.put("/{role_id}/permissions", response_model=Response[RoleSchema])
async def update_role_permissions(
    *,
    role_id: UUID,
    permissions_in: RolePermissionsUpdate,
    current_user: User = Depends(deps.PermissionChecker("user:manage")),
) -> Any:
    """
    Update role permissions (replace all).
    """
    try:
        role = await Role.get(id=role_id)
    except DoesNotExist:
        raise HTTPException(
            status_code=404,
            detail="Role not found",
        )
    
    if role.is_system_role:
        raise HTTPException(
            status_code=400,
            detail="System role permissions cannot be modified",
        )
    
    # Clear existing permissions
    await role.permissions.clear()
    
    # Add new permissions
    for perm_code in permissions_in.permissions:
        try:
            perm = await Permission.get(code=perm_code)
            await role.permissions.add(perm)
        except DoesNotExist:
            raise HTTPException(
                status_code=400,
                detail=f"Permission '{perm_code}' not found",
            )
    
    role = await Role.get(id=role_id).prefetch_related("permissions")
    return success(data=role, msg="Role permissions updated successfully")


@router.delete("/{role_id}", response_model=Response[RoleSchema])
async def delete_role(
    role_id: UUID,
    current_user: User = Depends(deps.PermissionChecker("user:manage")),
) -> Any:
    """
    Delete a role.
    """
    try:
        role = await Role.get(id=role_id).prefetch_related("permissions")
    except DoesNotExist:
        raise HTTPException(
            status_code=404,
            detail="Role not found",
        )
    
    if role.is_system_role:
        raise HTTPException(
            status_code=400,
            detail="System roles cannot be deleted",
        )
    
    # Check if role is assigned to any users
    users_with_role = await User.filter(roles=role).count()
    if users_with_role > 0:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot delete role: {users_with_role} user(s) are assigned to this role",
        )
    
    await role.delete()
    return success(data=role, msg="Role deleted successfully")

File: frontend/lib/api/roles.ts
import { api } from './client'
import type { PageData } from './users'

export interface Permission {
  id: string
  scope: string
  code: string
  description: string | null
}

export interface Role {
  id: string
  name: string
  description: string | null
  is_system_role: boolean
  permissions: Permission[]
}

export interface RoleCreateInput {
  name: string
  description?: string
  permissions?: string[]
}

export interface RoleUpdateInput {
  name?: string
  description?: string
}

export interface RolePermissionsUpdateInput {
  permissions: string[]
}

export interface PermissionCreateInput {
  scope: string
  code: string
  description?: string
}

export interface PermissionUpdateInput {
  scope: string
  code: string
  description?: string
}

export const rolesApi = {
  /**
   * 获取角色列表
   */
  getRoles: async (page: number = 1, pageSize: number = 50): Promise<PageData<Role>> => {
    return api.get<PageData<Role>>(`/roles?page=${page}&page_size=${pageSize}`)
  },

  /**
   * 获取单个角色
   */
  getRole: async (id: string): Promise<Role> => {
    return api.get<Role>(`/roles/${id}`)
  },

  /**
   * 创建角色
   */
  createRole: async (data: RoleCreateInput): Promise<Role> => {
    return api.post<Role>('/roles', data)
  },

  /**
   * 更新角色
   */
  updateRole: async (id: string, data: RoleUpdateInput): Promise<Role> => {
    return api.put<Role>(`/roles/${id}`, data)
  },

  /**
   * 更新角色权限
   */
  updateRolePermissions: async (id: string, permissions: string[]): Promise<Role> => {
    return api.put<Role>(`/roles/${id}/permissions`, { permissions })
  },

  /**
   * 删除角色
   */
  deleteRole: async (id: string): Promise<Role> => {
    return api.delete<Role>(`/roles/${id}`)
  },
}

export const permissionsApi = {
  /**
   * 获取权限列表
   */
  getPermissions: async (page: number = 1, pageSize: number = 100, scope?: string): Promise<PageData<Permission>> => {
    let url = `/permissions?page=${page}&page_size=${pageSize}`
    if (scope) url += `&scope=${scope}`
    return api.get<PageData<Permission>>(url)
  },

  /**
   * 获取单个权限
   */
  getPermission: async (id: string): Promise<Permission> => {
    return api.get<Permission>(`/permissions/${id}`)
  },

  /**
   * 创建权限
   */
  createPermission: async (data: PermissionCreateInput): Promise<Permission> => {
    return api.post<Permission>('/permissions', data)
  },

  /**
   * 更新权限
   */
  updatePermission: async (id: string, data: PermissionUpdateInput): Promise<Permission> => {
    return api.put<Permission>(`/permissions/${id}`, data)
  },

  /**
   * 删除权限
   */
  deletePermission: async (id: string): Promise<Permission> => {
    return api.delete<Permission>(`/permissions/${id}`)
  },
}
